File: src/Phan_01/attributes_methods.cpp
#include <iostream>
#include <fstream>
using namespace std;

class Animal
{
public:
	// declare attributes
	string name;
	int age;

	// declare methods
	// define methods inside class
	void description()
	{
		cout << name << " is " << age << " years old\n";
	}
	void speak(string sound);
};

// define methods outside class
void Animal::speak(string sound)
{
	cout << name << " says " << sound << endl;
}

// create a function that takes an Animal object as an argument
void describe(Animal &animal)
{
	animal.description();
}

// create a function that sum 2 numbers
int sum(int a, int b)
{
	return a + b;
}

// create a random array of n numbers
int *randomArray(int n)
{
	int *arr = new int[n];
	for (int i = 0; i < n; i++)
	{
		arr[i] = rand() % 100;
	}
	return arr;
}

// find max number of int array
int max(int *arr, int n)
{
	int max = arr[0];
	for (int i = 1; i < n; i++)
	{
		if (arr[i] > max)
		{
			max = arr[i];
		}
	}
	return max;
}

// create a function that change numbers to string
string *changeToString(int *arr, int n)
{
	string *str = new string[n];
	for (int i = 0; i < n; i++)
	{
		str[i] = to_string(arr[i]);
	}
	return str;
}

// create a function that change numbers to char array
char *changeToChar(int *arr, int n)
{
	char *ch = new char[n];
	for (int i = 0; i < n; i++)
	{
		ch[i] = arr[i];
	}
	return ch;
}

// write a number to a file
void writeToFile(int n)
{
	ofstream file;
	file.open("number.txt");
	file << n;
	file.close();
}

int main()
{
	Animal animal;

	int n;
	cin >> n;

	animal.name = "rocky";
	animal.age = 10;

	animal.description();
	animal.speak("Grr");

	return 0;
}

File: src/Phan_01/attributes_methods.py
class Animal:
    def __init__(self, name, age):
        self.name = name
        self.age = age

    # declare and define methods inside class
    def description(self):
        return f"{self.name} is {self.age} years old"


# declare and define methods outside class
def func(self, sound):
    return f"{self.name} says {sound}"


Animal.speak = func

animal = Animal("rocky", 10)

print(animal.description())  # rocky is 10 years old
print(animal.speak("Grr"))  # rocky says Grr


# class Animal:
#     name = ""

#     def description(self):
#         # use self. to access attributes
#         return f"{self.name} is {self.age} years old"


# Animal.age = 0

# animal = Animal()
# animal.species = "dog"

# print(animal.species)  # dog
# # print(Animal.species)   # error